File: 8. Heaps/347. Top K Frequent Elements/Solution.cpp
#include <vector>
#include <unordered_map>
#include <queue>

class Solution {
public:
    std::vector<int> topKFrequent(std::vector<int>& nums, int k) {
        std::unordered_map<int, int> freq_map;

        for (auto const& num : nums)
        {
            ++freq_map[num];
        }

        typedef std::pair<int, int> pair;
        std::priority_queue<pair, std::vector<pair>, std::greater<pair>> min_heap;

        for (auto const& [val, freq] : freq_map)
        {
            min_heap.push({freq, val});

            if (min_heap.size() > k)
            {
                min_heap.pop();
            }
        }

        std::vector<int> res;

        while (!min_heap.empty())
        {
            res.push_back(min_heap.top().second);
            min_heap.pop();
        }

        return res;

    }
};

File: 8. Heaps/347. Top K Frequent Elements/Solution.py
from typing import List
from collections import Counter
import heapq

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq_map = Counter(nums)
        min_heap = []

        for val, freq in freq_map.items():
            heapq.heappush(min_heap, (freq, val))

            if len(min_heap) > k:
                heapq.heappop(min_heap)

        res = [heapq.heappop(min_heap)[1] for _ in range(k)]
        return res
                